### ISSUE TITLE
Regression 1.91 -> 1.92: false positive unused_assignments on struct field write when raw pointer exists

### ISSUE THREAD

**Issue by wmmc88** [rust-lang/rust#154483]
### Code

I tried this code:

```rust
struct S {
    value: u32,
}

fn main() {
    let mut s = S { value: 0 };
    let ptr = &raw mut s;
    s.value = 42;
    println!("{}", unsafe { (*ptr).value });
}
```

I expected to see this happen: no warnings, since `s.value` is read through `ptr` on the next line.

Instead, this happened:

```
warning: value assigned to `s` is never read
 --> repro.rs:8:5
  |
8 |     s.value = 42;
  |     ^^^^^^^^^^^^
  |
  = help: maybe it is overwritten before being read?
  = note: `#[warn(unused_assignments)]` (part of `#[warn(unused)]`) on by default

warning: 1 warning emitted
```

The program compiles and prints `42`, confirming the assignment is read through the raw pointer.

### Version it worked on

It most recently worked on: Rust 1.91.0

### Version with regression

`rustc --version --verbose`:
```
rustc 1.92.0 (ded5c06cf 2025-12-08)
binary: rustc
commit-hash: ded5c06cf21d2b93bffd5d884aa6e96934ee4234
commit-date: 2025-12-08
host: x86_64-pc-windows-msvc
release: 1.92.0
LLVM version: 21.1.3
```

### Backtrace

N/A — no crash, just a false positive lint warning.

### Additional context

The false positive specifically affects **struct field writes** when a raw pointer to the struct has been taken via `&raw mut`. Direct variable writes (e.g., `a = 42`) have always triggered this lint in this scenario, but struct field writes (e.g., `s.value = 42`) were correctly suppressed through 1.91 and started falsely firing in 1.92.

@rustbot modify labels: +regression-from-stable-to-stable -regression-untriaged
